### DISCUSSION TITLE
Can't get negate to work

### DISCUSSION THREAD

**Discussion by goyalyashpal** [ImageMagick/ImageMagick#5591] [Help] [answered]
Hi!
* I have just started to use image magick. 
* i have tried several combination of command `+negate` (trying to invert a grayscale image) 😅 but none seem to be working, 
* the output is just a non-zero sized png file which looks black on thumbnail but ain't opening with any viewers

combos tried:
* `magick convert -negate trial.png out.png` (wrong as per https://imagemagick.org/Usage/basics/#syntax )
* `magick trial.png -negate trialn.png`
* `magick trial.png +negate trialn.png`
* `convert trial.png -negate trail.png`
* `convert trial.png +negate trail.png`
* `magick convert trial.png -negate out.png`

now i just can't figure out what is going wrong 😅 

### Environment:
* `magick --version` ImageMagick 7.1.0-49 Q16-HDRI x64 7a3f3f1:20220924 
* Windows 10 20H2
* bash available via msys2

**Comment by GeeMack** [comment, marked as answer]
First, in almost every case when using ImageMagick version 7 it's better to use just "magick" instead of "magick convert".  Adding "convert" forces IMv7 to use IM version 6 behavior. By omitting "convert" the command will act with proper IMv7 behavior. There are notable differences in the way each version handles the alpha channel.

Using `-negate`  acts on all the channels, even the alpha channel if it exists. To get `-negate` to act on only the red, green, and blue channels, try this...

    magick trial.png -channel RGB -negate +channel trialn.png

Setting that `-channel` before negating specifies the following operation(s) are only to be applied to the RGB channels. After negating, add `+channel` to make sure any following operations are applied to all channels.

**Reply by goyalyashpal** [reply]
0. Thanks a lot, yes, that worked :)
1. What does `+negate` do?
2. I think the documentation can be improved to better reflect this fact. Currently, the documentation lists negate at several place, and above thing could be hinted at some of them to facilitate such knowledge

* at [clp#operator]

> In this example, -negate negates the wand image but not the wizard:
> `magick wand.png -negate wizard.png images.png`

* documentation at [clo#negate]

> -negate
> Replace each pixel with its complementary color.
>
> The red, green, and blue intensities of an image are negated. White becomes black, yellow becomes blue, etc. Use [+negate](https://imagemagick.org/script/command-line-options.php#negate) to only negate the grayscale pixels of the image.


[clo#negate]: https://imagemagick.org/script/command-line-options.php#negate
[clp#operator]: https://imagemagick.org/script/command-line-processing.php#operator


3. Unrelated to negate, but related to documentation, maybe there should be some page on the correspondence/mapping betweeen the local site supplied with magick and the online version:
    + local: `www/---.html#---`
    + online: `script/---.php#---`

**Reply by goyalyashpal** [reply]
Other related questions (saving for my own reference:)

* Swapphing RB channels: https://github.com/ImageMagick/ImageMagick/discussions/2160#discussioncomment-27235
* Inverting in bulk from a folder: https://github.com/ImageMagick/ImageMagick/discussions/2124#discussioncomment-24084
* camera image : scanning like filter: https://github.com/ImageMagick/ImageMagick/discussions/4145#discussioncomment-1272563
* working with layers: https://github.com/ImageMagick/ImageMagick/discussions/3319#discussioncomment-414603

One more, though 'ts resolved already:
* Difference in negate in IM7 and IM6: https://github.com/ImageMagick/ImageMagick/discussions/2124#discussioncomment-3725254

**Comment by snibgo** [comment]
As GeeMack says, "`-negate`" operates on all active channels. In v7, if an image has alpha then it is active unless we use "-channel". The same is true of other operations, eg "`-evaluate divide 2`" will by default also divide alpha by 2.

"`+negate`" will negate only pixels that are some shade of gray, including black and white. Other pixels (pixels that have a colour) are unchanged. For example:
```
magick logo: +negate n.png
```
![n](https://user-images.githubusercontent.com/33812368/192120410-f4492f2b-bf2b-41dc-bf69-e1b2df1ad38e.png)
I don't think I have ever wanted to do this.

**Reply by goyalyashpal** [reply]
> I don't think I have ever wanted to do this.

hehe, this can be helpful when the image you are dealing with are already in grayscale

**Reply by goyalyashpal** [reply]
but yeah, the point being, even this didn't work on those, and faced similar fate to `-negate`

**Reply by fmw42** [reply]
Even what did not work. Please explain what you need to do and you exact command line.

**Comment by snibgo** [comment]
@yashpalgoyal1304: Your sample command ...
```
magick trial.png -negate trialn.png
```
... should work. If it doesn't, please show your input and output files.

**Reply by goyalyashpal** [reply]
sorry @ snibgo and @ fmw42 for the late reply.

### Reply to/Quoting:

> Even what did not work. Please explain what you need to do and you exact command line.

> Your sample command should work. If it doesn't, please show your input and output files.


### Reproduction steps:
_I am hiding the inline image views here, open these images in new tab etc:_

* Input file: [trial]
    + yes, have confirmed that this uploaded file behaves same as my local file for both the following mentioned commands
+ `magick trial.png -negate trialn.png` \
output [BAD] for above command (was my own try): [trialn] 
    + yes, it is blank - tested in opening in browser, picasa viewer, GIMP, irfan viewer, win 10 photos application, etc.
    + the only place it shows up is the ~~thumbnails~~ sidebar/preview pane in windows 10 explorer - see the below attached image (with inline view enabled)
- `magick trial.png -channel RGB -negate +channel trialn.png` \
output [GOOD] for above command (given by @ GeeMack) : [works]

[trial]: https://user-images.githubusercontent.com/19423063/197410551-026a6dae-6b7d-4610-b1b3-988d6d911073.png
[trialn]: https://user-images.githubusercontent.com/19423063/197410640-31127b18-cf8c-400f-a001-775f88c17c94.png
[works]: https://user-images.githubusercontent.com/19423063/197410825-f6b1e994-9263-45ca-b247-6c50d635c2d7.png

### Env:
* magick: ImageMagick 7.1.0-49 Q16-HDRI x64 7a3f3f1:20220924
* os: win 10 20h2

### Other screenshots
the bad output image shown in the preview pane, but blank everywhere else (thumbnails included)

![image](https://user-images.githubusercontent.com/19423063/197411579-5b8d014c-a98e-4a1d-b9b5-4e3a7a39504a.png)

**Comment by snibgo** [comment]
As GeeMack says, "-negate" operates on all channels, including alpha if present. Your image has an alpha channel. Use "-channel RGB" to prevent alpha from being negated.

**Reply by goyalyashpal** [reply]
> Your image has an alpha channel.

1. how can i check about that i.e. if an image has alpha channel & possible some preview of it?
2. how to remove alpha channel in pre?
3. how to remove alpha channel in post?

**Comment by snibgo** [comment]
```
magick yourfile.ext -format %r info:
```
... will tell you if the image has alpha. For example:

```
C:\web\im>%IMG7%magick toes_holed.png -format %r info:
DirectClass sRGB Alpha
```

You can visualise the alpha channel like this:
```
magick yourfile.ext -alpha extract x.png
```
... then view x.png. It will be blacck where alpha was zero.

To turn alpha off:
```
magick yourfile.ext -alpha off x.png
```

I don't know what you mean by "in pre" and "in post".

**Reply by goyalyashpal** [reply]
with pre- i kinda meant that removing alpha channel before applying other operations on it in `magick`, and with post, i meant applying the alpha channel removal operation in the very end after all other operations

**Comment by snibgo** [comment]
Using IM v7 with "magick", operations are performed in the order they appear in the command. So you can "-alpha off" before or after any operations you like.
